### DISCUSSION TITLE
Optional query params are broken in swagger when pydantic2 is installed

### DISCUSSION THREAD

**Discussion by zoola969** [fastapi/fastapi#10836] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def root(
    q: int | None = Query(None),
):
    return q
```

### Description

Open `/docs` in browser. I expect optional query parameter with integer type, but there is just common text optional query param
pydantic v2
![image](https://github.com/tiangolo/fastapi/assets/34693850/a636bc4d-31a0-44dc-849d-85cc4a2d4543)
pydantic v1
![image](https://github.com/tiangolo/fastapi/assets/34693850/ec35003f-bb39-4e02-8a9e-b46c5a220aa4)

### Operating System

Linux

### FastAPI Version

0.105.0

### Pydantic Version

2.5.3

### Python Version

3.11.7

**Comment by JavierSanchezCastro** [comment]
Use Annotated like

```python3
from fastapi import FastAPI, Query
from typing import Annotated

app = FastAPI()


@app.get("/")
def root(q: Annotated[int, Query()] = None):
    return q
```

![imagen](https://github.com/tiangolo/fastapi/assets/72013291/32252754-4ed6-46f1-a27e-7ba41b4ebc75)

**Reply by zoola969** [reply]
It works in swagger, but this code has incorrect typing because `q` should have a type `int | None`, not `int`.

**Reply by JavierSanchezCastro** [reply]
Look at this: https://github.com/tiangolo/fastapi/discussions/10781#discussioncomment-7856305

As that comment says, it depends on how those pages choose to render the schema.

**Reply by zoola969** [reply]
It didn't help me to understand how to correctly define optional query param.
I look at `openapi.json` data and see the difference in how the schema is generated after switching to pydantic v2
p1: 
"parameters":[{"required":false,"schema":{"type":"integer","title":"Q"},"name":"q","in":"query"}]
p2:
"parameters":[{"name":"q","in":"query","required":false,"schema":{"anyOf":[{"type":"integer"},{"type":"null"}],"title":"Q"}}]

So, there is definitely a bug in fastapi or pydantic

**Reply by JavierSanchezCastro** [reply]
https://github.com/tiangolo/fastapi/discussions/10781#discussioncomment-7887787

The problem with p1 lies in its allowance of a null value, which is not reflected in the openapi.json. P1 represents the oldest openapi generated schema, and Swagger is capable of displaying this information because there is only one type for the query parameter in the generated schema, even though it permits a null value.

However, with the corrected openapi schema as p2, there are now two types for the same parameter. I suppose Swagger doesn't know how to display that information, but you can see that in /redoc that query parameter would be an integer or null.

That's what I mean by it depends on how those pages decide to render the schema

**Reply by zoola969** [reply, marked as answer]
Thanks, I got it. And it helps me to find several issues about it and one hack solution. 

Using `SkipJsonSchema` removes anyOf and null from openapi.json and swagger shows parameter correctly
```python
q: Annotated[int | SkipJsonSchema[None], Query()] = None
```
`"parameters":[{"name":"q","in":"query","required":false,"schema":{"type":"integer","title":"Q"}}]`

https://github.com/tiangolo/fastapi/discussions/9709#discussioncomment-6437341
https://github.com/pydantic/pydantic/pull/6653#issuecomment-1646211152

It seems like swagger doesn't support anyOf for parameters

**Reply by zoola969** [reply]
btw will be nice to see this solution or problem description somewhere in docs, this is a stopper of using p2 for me

**Reply by leonidktoto** [reply]
@tiangolo `q: Annotated[int | SkipJsonSchema[None], Query()] = None` I started studying fastapi, I think this solution is quite working. what do you say? Is there another solution to this problem?

**Comment by sabonerune** [comment]
@tiangolo 
By the way, are there any plans to fix this problem in FastAPI?

https://json-schema.org/understanding-json-schema/reference/null
>It's important to remember that in JSON, null isn't equivalent to something being absent.

I don't think FastAPI(or maybe OpenAPI) has a way to accept null in queries.

**Comment by maurerle** [comment]
So I came across this issue from the docs here:
https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#__tabbed_15_1

which say that:
```python
from typing import Annotated

from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/items/")
async def read_items(q: Annotated[list[str] | None, Query()] = None):
    query_items = {"q": q}
    return query_items
```

should produce an array in Swagger UI/OpenAPI, which did not happen.
Solutions to this are either:

remove the optional "None" type hint:

```python
from typing import Annotated

from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/items/")
async def read_items(q: Annotated[list[str], Query()] = None):
    query_items = {"q": q}
    return query_items
```
or downgrade pydantic:

`pip install "pydantic<2"`

However the current docs situation is quite bad for this, even if this is not a FastAPI issue originally.

Using Redoc UI at `/redoc` does not have a tryout mode, but works fine using the `anyOf` param.
The problem is, that the current swagger UI does not work with `anyOf` param, even if the second anyOf is of type:null `"anyOf":[{"type":"string","minLength":3,"maxLength":50},{"type":"null"}]` (which is redundant information anyway, as `required: false` is given too.

**Reply by maurerle** [reply]
There is also https://github.com/tiangolo/fastapi/discussions/10481 as a similar discussion.

In Pydantic, there is this issue: https://github.com/pydantic/pydantic/issues/6647 which resulted in an additional wrapper:
https://github.com/pydantic/pydantic/pull/6653 which can be used as

```python
from pydantic import SkipJsonSchema

@app.get("/items/")
async def read_items(q: Annotated[list[str] | SkipJsonSchema[None], Query()] = None):
    query_items = {"q": q}
    return query_items
```

which is not very well either.

Swagger somehow seems to support `anyOf` argument as seen here: https://github.com/swagger-api/swagger-ui/issues/9056 but somehow fails for our use case.

There is also as FastAPI related PR in https://github.com/tiangolo/fastapi/pull/9873
which can be installed using
`python3 -m pip install --force-reinstall git+https://github.com/tiangolo/fastapi/@feat/custom-generate-json-schema`
It however produces
`AttributeError: property 'mode' of 'GenerateJsonSchema' object has no setter`
as it does not seem up to date.
